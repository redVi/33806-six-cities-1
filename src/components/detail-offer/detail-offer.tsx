import React, {Component} from 'react';
import {connect} from 'react-redux';

import {OfferType, CommentType} from '@/types';
import {redirectToId} from '@/helpers';
import Comments from '@/api/comments';
import {dataActionCreator} from '@/reducer/data/data';
import {checkAuthorization} from '@/reducer/user/selectors';
import {getComments, getOfferById, getSimilarOffers} from '@/reducer/data/selectors';

import Mark from '@/components/mark/mark';
import Bookmark from '@/components/bookmark/bookmark';
import CityMap from '@/components/city-map/city-map';
import PlacesList from '@/components/places-list/places-list';
import Rating from '@/components/rating/rating';
import Reviews from '@/components/reviews/reviews';
import Host from '@/components/host/host';
import Inside from '@/components/inside/inside';
import Price from '@/components/price/price';
import Features from '@/components/features/features';
import Gallery from '@/components/gallery/gallery';

interface Props {
  history?: any[],
  id: number,
  offer: OfferType,
  offers: OfferType[],
  comments: CommentType[],
  isLoggedIn: boolean,
  getComments: () => object[],
  setActiveItem?: () => void
}

class DetailOffer extends Component<Props> {
  componentDidMount(): void {
    this.props.getComments();
  }

  componentDidUpdate(prevProps: Readonly<Props>): void {
    if (this.props.id !== prevProps.id) {
      this.props.getComments();
    }
  }

  render() {
    const {offer, offers, comments, isLoggedIn} = this.props;
    const redirect = (item) => redirectToId(`/offer/${item.id}`, this.props.history);

    if (!offer) return null;

    return (
      <div className="page">
        <main className="page__main page__main--property">
          <section className="property">

            <Gallery images={offer.images} maxCount={6} />

            <div className="property__container container">
              <div className="property__wrapper">

                {offer.isPremium ? <Mark className="property__mark" /> : null}

                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {offer.title}
                  </h1>
                </div>

                <Bookmark
                  id={offer.id}
                  isFavorite={offer.isFavorite}
                  className="property" width="31" height="33" />

                <Rating className="property" rating={offer.rating}>
                  <span className="property__rating-value rating__value">
                    {offer.rating}
                  </span>
                </Rating>

                <Features
                  type={offer.type}
                  bedrooms={offer.bedrooms}
                  maxAdults={offer.maxAdults} />

                <Price price={offer.price} className="property__price" text="&nbsp;night" />

                <Inside items={offer.goods} title="What&apos;s inside" />

                <Host host={offer.host} />

                <Reviews
                  id={offer.id}
                  maxCount={10}
                  isLoggedIn={isLoggedIn}
                  comments={comments} />
              </div>
            </div>

            <CityMap
              location={offer.location}
              items={offers.concat(offer)}
              hasSelectedItem={true}
              className="property__map" />

          </section>

          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <PlacesList places={offers} className="near-places__list" onImageClick={redirect} />
            </section>
          </div>
        </main>
      </div>
    );
  }
}

const mapStateToProps = (state: object, ownProps) => {
  const id = ownProps.match.params.id;

  return Object.assign({}, ownProps, {
    id: id,
    offer: getOfferById(state, id),
    offers: getSimilarOffers(state, id).slice(0, 3),
    comments: getComments(state),
    isLoggedIn: checkAuthorization(state)
  });
};

const mapDispatchToProps = (dispatch: Function, ownProps) => ({
  getComments: () => {
    Comments.get(ownProps.match.params.id).then((response) => {
      dispatch(dataActionCreator.fetchComments(response.data));
    });
  }
});

export {DetailOffer};
export default connect(mapStateToProps, mapDispatchToProps)(DetailOffer);
