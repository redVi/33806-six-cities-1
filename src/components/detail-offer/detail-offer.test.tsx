import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import {DetailOffer} from "@/components/detail-offer/detail-offer";

const renderer = new ShallowRenderer();

const offer = {
  id: 24,
  title: "Hotel",
  price: 120,
  rating: 4.6,
  images: ["house.jpg", "house2.jpg"],
  goods: ["Breakfast", "Towels"],
  bedrooms: 3,
  maxAdults: 2,
  isFavorite: false,
  host: {
    name: "Bob",
    avatarUrl: "bob\'s-avarar.jpg",
    isPro: false,
    description: "The best host in the world"
  }
};

const comment = {
  id: 1,
  rating: 5,
  comment: "Nice place",
  date: "2019-06-10T02:34:08.739Z",
  user: {
    name: "Tina",
    avatarUrl: "tina-avatar.jpg"
  }
};

describe("DetailOffer", () => {
  it("renders correctly", () => {
    renderer.render(
      <DetailOffer
        id={24}
        offer={offer}
        offers={[offer]}
        comments={[comment]}
        isLoggedIn={false}
        getComments={jest.fn()}
      />
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });
});
