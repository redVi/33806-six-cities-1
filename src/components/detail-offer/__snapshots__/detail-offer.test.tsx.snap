// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailOffer renders correctly 1`] = `
<div
  className="page"
>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <Gallery
        images={
          Array [
            "house.jpg",
            "house2.jpg",
          ]
        }
        maxCount={6}
      />
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Hotel
            </h1>
          </div>
          <withRouter(withProps)
            className="property"
            height="33"
            id={24}
            isFavorite={false}
            width="31"
          />
          <Rating
            className="property"
            rating={4.6}
          >
            <span
              className="property__rating-value rating__value"
            >
              4.6
            </span>
          </Rating>
          <Features
            bedrooms={3}
            maxAdults={2}
          />
          <Price
            className="property__price"
            price={120}
            text="Â night"
          />
          <Inside
            items={
              Array [
                "Breakfast",
                "Towels",
              ]
            }
            title="What's inside"
          />
          <Host
            host={
              Object {
                "avatarUrl": "bob's-avarar.jpg",
                "description": "The best host in the world",
                "isPro": false,
                "name": "Bob",
              }
            }
          />
          <Reviews
            comments={
              Array [
                Object {
                  "comment": "Nice place",
                  "date": "2019-06-10T02:34:08.739Z",
                  "id": 1,
                  "rating": 5,
                  "user": Object {
                    "avatarUrl": "tina-avatar.jpg",
                    "name": "Tina",
                  },
                },
              ]
            }
            id={24}
            isLoggedIn={false}
            maxCount={10}
          />
        </div>
      </div>
      <CityMap
        className="property__map"
        hasSelectedItem={true}
        items={
          Array [
            Object {
              "bedrooms": 3,
              "goods": Array [
                "Breakfast",
                "Towels",
              ],
              "host": Object {
                "avatarUrl": "bob's-avarar.jpg",
                "description": "The best host in the world",
                "isPro": false,
                "name": "Bob",
              },
              "id": 24,
              "images": Array [
                "house.jpg",
                "house2.jpg",
              ],
              "isFavorite": false,
              "maxAdults": 2,
              "price": 120,
              "rating": 4.6,
              "title": "Hotel",
            },
            Object {
              "bedrooms": 3,
              "goods": Array [
                "Breakfast",
                "Towels",
              ],
              "host": Object {
                "avatarUrl": "bob's-avarar.jpg",
                "description": "The best host in the world",
                "isPro": false,
                "name": "Bob",
              },
              "id": 24,
              "images": Array [
                "house.jpg",
                "house2.jpg",
              ],
              "isFavorite": false,
              "maxAdults": 2,
              "price": 120,
              "rating": 4.6,
              "title": "Hotel",
            },
          ]
        }
      />
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <PlacesList
          className="near-places__list"
          onImageClick={[Function]}
          places={
            Array [
              Object {
                "bedrooms": 3,
                "goods": Array [
                  "Breakfast",
                  "Towels",
                ],
                "host": Object {
                  "avatarUrl": "bob's-avarar.jpg",
                  "description": "The best host in the world",
                  "isPro": false,
                  "name": "Bob",
                },
                "id": 24,
                "images": Array [
                  "house.jpg",
                  "house2.jpg",
                ],
                "isFavorite": false,
                "maxAdults": 2,
                "price": 120,
                "rating": 4.6,
                "title": "Hotel",
              },
            ]
          }
        />
      </section>
    </div>
  </main>
</div>
`;
